src/connector/src/parser/additional_columns.rs

```diff
@@ -0,0 +1,179 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashMap;
+use std::sync::LazyLock;
+
+use risingwave_common::catalog::{ColumnCatalog, ColumnDesc, ColumnId};
+use risingwave_common::types::DataType;
+use risingwave_pb::plan_common::AdditionalColumnType;
+
+use crate::source::{
+    KAFKA_CONNECTOR, KINESIS_CONNECTOR, PULSAR_CONNECTOR, S3_CONNECTOR, S3_V2_CONNECTOR,
+};
+
+pub type CompatibleAdditionalColumnsFn =
+    Box<dyn Fn(ColumnId, &str) -> ColumnCatalog + Send + Sync + 'static>;
+
+pub static CONNECTOR_COMPATIBLE_ADDITIONAL_COLUMNS: LazyLock<
```

Comment:
Hmm... not a big deal, but a function looks more straight-forward to me:

```rust
pub fn get_additional_column_catalog(connector_name: &str, column_name: &str, column_id: &ColumnId) -> Result<ColumnCatalog>
```

Reply:
it is following the practice of `CONNECTORS_COMPATIBLE_FORMATS`
https://github.com/singularity-data/risingwave/blob/30f95165160479b39638fd226bb58086741efc38/src/frontend/src/handler/create_source.rs#L965-L1031

Reply:
I see. Also don't really understand why it prefers `HashMap` instead of `match-case` (which implies writing a function instead of `map`)

Reply:
No specific reason. Just following the practice. 
If you mean `CONNECTORS_COMPATIBLE_FORMATS `, it is because the each connector has its accepted format and encode. The combination is too complex for a match.

Reply:
Just read the code of `CONNECTORS_COMPATIBLE_FORMATS`. It's more sensible to me because it holds **data** instead of function references. 

From my understanding of Rust, here I prefer to use simple pattern match instead of hashmap.